src/iter/collect/consumer.rs

```diff
@@ -108,6 +136,26 @@ impl<'c, T: Send + 'c> UnindexedConsumer<T> for CollectConsumer<'c, T> {
         unreachable!("CollectConsumer must be indexed!")
     }
     fn to_reducer(&self) -> Self::Reducer {
-        NoopReducer
+        CollectReducer
+    }
+}
+
+/// CollectReducer combines adjacent chunks; the result must always
+/// be contiguous so that it is one combined slice.
+pub(super) struct CollectReducer;
+
+impl<'c, T> Reducer<CollectResult<'c, T>> for CollectReducer {
+    fn reduce(
+        self,
+        mut left: CollectResult<'c, T>,
+        right: CollectResult<'c, T>,
+    ) -> CollectResult<'c, T> {
+        // Merge if the CollectResults are adjacent and in left to right order
+        // else: drop the right piece now and total length will end up short in the end,
+        // when the correctness of the collected result is asserted.
+        if left.start.wrapping_add(left.len) == right.start {
```

Comment:
Why did you choose `wrapping_add` here?

Also, I worry a little that this can't track ZSTs properly, but I guess the aggregate result will still drop items if it comes up short of the final length, right?

Reply:
`wrapping_add` was chosen just since it's a safe method, and it seems to do the job, it's more or less an arithmetic check. It's not used for wrapping, but because the requirements compared with `add` are somewhat relaxed. 

Right, we don't track ZST the same way.. we can't see if they are adjacent or not, and just add up the lengths; so we could totally be fooled into merging "non-adjacent" chunks. We add up the length and will drop the right number of them in the end.